9.0 Port account_cutoff_base and account_cutoff_prepaid

alexis-via opened the pull request:
Extract a new module account_invoice_start_end_dates from account_cutoff_prepaid

Warning: it is not a port of the OCA/account-closing/8.0 version of the module, but a port of my LP branch https://code.launchpad.net/~akretion-team/account-closing/80-forecast-prepaid which is still the branch I use for Odoo v8 (which explains why I wasn't active on OCA/account-closing so far). This is due to the fact that my "forecast" branches were not merged on LP before the transition to github.

So the next step is to merge the few enhancements that took place in the OCA version of the module.

alexis-via committed:
Port account_cutoff_base and account_cutoff_prepaid to Odoo v9

Extract a new module account_invoice_start_end_dates from account_cutoff_prepaid

alexis-via commented:
And of course, this PR is also a port to the new API.

Commit:
[ADD] Add setup.py

Commit:
[ADD] Add some hooks that are in 8.0

alexis-via commented:
@adrienpeiffer I don't see the pull request you told me about...

Commit:
Merge pull request #1 from acsone/9.0-cutoff-setup-ape

[ADD] Add some hooks that are in 8.0

adrienpeiffer reviewed (approved):
:+1: Code review and functional test

adrienpeiffer commented:
@chrisdec do you have a moment to test it on runbot ? :smile:

gurneyalex reviewed (approved):
code review, no test.

pedrobaeza commented on account_cutoff_base/models/account_cutoff.py:
```diff
@@ -0,0 +1,359 @@
+
+    @api.model
+    def _inherit_default_cutoff_account_id(self):
+        '''Function designed to be inherited by other cutoff modules'''
```
Docstrings should use double triple-quota (applicable to all)

alexis-via replied:
If docstring should really use double triple-quota, then why flake8/pylint says nothing about it ?

pedrobaeza replied:
Yes, good point. We need to add that check. But you can see here that what I say is correct: https://www.python.org/dev/peps/pep-0257/#id15 (`For consistency, always use """triple double quotes""" around docstrings. Use r"""raw triple double quotes""" if you use any backslashes in your docstrings. For Unicode docstrings, use u"""Unicode triple-quoted strings""" .`)

pedrobaeza commented on account_cutoff_base/models/account_cutoff.py:
```diff
@@ -0,0 +1,359 @@
+            movelines_to_create.append((0, 0, vals))
+            amount_total += amount
+
+        # add contre-partie
```
counter-part

pedrobaeza commented on account_cutoff_base/security/account_cutoff_base_security.xml:
```diff
@@ -1,11 +1,20 @@
-    </data>
-</openerp>
+<odoo>
+<data noupdate="1">
```
Put directly `<odoo noupdate="1">`

pedrobaeza commented on account_cutoff_base/views/account_cutoff.xml:
```diff
@@ -1,30 +1,28 @@
 
-<openerp>
+<odoo>
 <data>
```
You can remove `<data>` tag

pedrobaeza commented on account_cutoff_base/views/account_cutoff.xml:
```diff
@@ -35,15 +33,15 @@
-                        <field name="cutoff_journal_id"/>
-                        <field name="cutoff_account_id"/>
-                        <field name="move_label"/>
+                        <field name="cutoff_journal_id" required="1"/>
```
Shouldn't this be put (required flag) at field level?

alexis-via replied:
It is set at view level because it is inherited in the module account_cutoff_prepaid and replaced by attrs = {'required': [('forecast', '=', False)]}

pedrobaeza replied:
OK, thanks for the explanation. A comment in the XML would save this question for future readers of the code.

pedrobaeza commented on account_cutoff_prepaid/i18n/account_cutoff_prepaid.pot:
```diff
@@ -6,8 +6,8 @@ msgid ""
 msgstr ""
```
Remove POT file

pedrobaeza commented on account_invoice_start_end_dates/demo/product_demo.xml:
```diff
@@ -0,0 +1,34 @@
+-->
+
+<odoo>
+<data>
```
Remove `<data>` tag

pedrobaeza commented on account_invoice_start_end_dates/models/account_invoice.py:
```diff
@@ -0,0 +1,91 @@
+    _inherit = 'account.invoice'
+
+    def inv_line_characteristic_hashcode(self, invoice_line):
+        '''Add start and end dates to hashcode used when the option "Group
```
Double quote

pedrobaeza reviewed (changes requested):
Several change across full code

chrisdec commented:
Hello, 

Functional test : 

Just one note : is it possible to show the message "Don't forget to Re-Generate Lines after entering or leaving forecast mode." only when user actives or disables the checkbox "forecast"? It's quite confusing to have this message when you open a new prepaid expense or a new prepaid revenue.

Except that, 👍

alexis-via committed:
Take into account the remarks of Pedro:

Use triple double quotes for docstring
Replace <openerp> by <odoo> in XML
Remove <data> tags in XML
Remove POT files

alexis-via commented:
@chrisdec Yes, we need to find a better way. I guess a better way would be to use a button to enter/leave forecast mode (like the "Active" button on partners for example) and delete all lines when the user clicks on this button. I'll try to find time to do that...
